### PR TITLE
Give better hint for YAML issues when building the reference index

### LINKED ISSUE #1351: build_reference will fail if a function is named N
I have a legacy function `N` in my package. While this name is not a good one, it is not what I usually consider as a "weirdly named function, e.g. "`+.gg`"" (as in [the doc](https://pkgdown.r-lib.org/reference/build_reference.html)).

Still, I have to escape it for `build_reference` to work:

```yaml
reference:
- title: "Won't work"
- contents: 
  - N
```

```yaml
reference:
- title: "Will work"
- contents: 
  - "N"
```

Here is the log:

```console
> build_reference()
-- Building function reference -------------------------------------------------
Error in FUN(X[[i]], ...) : invalid first argument
> traceback()
23: FUN(X[[i]], ...)
22: vapply(.x, .f, .mold, ..., USE.NAMES = FALSE)
21: map_mold(.x, .f, logical(1), ...)
20: map_lgl(nms, exists, envir = env, inherits = inherit)
19: env_has(env, string)
18: .f(.x[[i]], ...)
17: purrr::map(match_strings, match_eval, env = match_env(topics))
16: select_topics(section$contents, pkg$topics)
15: .f(.x[[i]], ...)
14: purrr::map(., data_reference_index_rows, pkg = pkg)
13: function_list[[i]](value)
12: freduce(value, `_function_list`)
11: `_fseq`(`_lhs`)
10: eval(quote(`_fseq`(`_lhs`)), env, env)
9: eval(quote(`_fseq`(`_lhs`)), env, env)
8: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
7: meta %>% purrr::map(data_reference_index_rows, pkg = pkg) %>% 
       purrr::compact() %>% unlist(recursive = FALSE)
6: data_reference_index(pkg)
5: stopifnot(is.list(x), is.list(val))
4: utils::modifyList(data, data_template(pkg, depth = depth))
3: render_page(pkg, "reference-index", data = data_reference_index(pkg), 
       path = "reference/index.html")
2: build_reference_index(pkg)
1: build_reference()
```

While the error may be justified, I think the error message might deserve to be clearer.

### PR DESCRIPTION
**Pull request by maelle**
Fix #1351

### CODE REVIEW

File: R/build-reference.R [hunk 1]
```diff
@@ -27,10 +27,10 @@
 #' reference:
 #' - title: Build
 #'   desc:  Build a complete site or its individual section components.
-#' - contents:
+#'   contents:
```
Comment (maelle) [inline R/build-reference.R:30]: also fixes docs

**Comment by maelle** [comment]
This will probably get a bit more complex after #1251

File: R/build-reference-index.R [hunk 2]
```diff
@@ -42,6 +42,23 @@ data_reference_index_rows <- function(section, pkg) {
 
 
   if (has_name(section, "contents")) {
+    if (any(!is.character(section$contents) || any(!nzchar(section$contents)))) {
+      abort(
```
Comment (hadley) [inline R/build-reference-index.R:46]: Can you please reduce the deep nesting with some intermediate variables?

File: R/build-reference-index.R [hunk 3]
```diff
@@ -42,6 +42,23 @@ data_reference_index_rows <- function(section, pkg) {
 
 
   if (has_name(section, "contents")) {
+    if (any(!is.character(section$contents) || any(!nzchar(section$contents)))) {
```
Comment (hadley) [inline R/build-reference-index.R:45]: What is the type of `section$contents`?

**Comment by maelle** [comment]
Putting this in the PR where I'll add external topics, as it changes code a lot.

**Pull request closed without merging**